payjoin/src/send/v2/mod.rs

```diff
@@ -344,9 +351,96 @@ mod test {
             },
             reply_key: HpkeKeyPair::gen_keypair().0,
         };
-        let serialized = serde_json::to_string(&req_ctx)?;
+        sender.v1.endpoint.set_exp(SystemTime::now() + Duration::from_secs(60));
+        sender.v1.endpoint.set_receiver_pubkey(HpkeKeyPair::gen_keypair().1);
+        sender.v1.endpoint.set_ohttp(OhttpKeys(
+            ohttp::KeyConfig::new(KEY_ID, KEM, Vec::from(SYMMETRIC)).expect("valid key config"),
+        ));
+
+        Ok(sender)
+    }
+
+    #[test]
+    fn sender_ser_de_roundtrip() -> Result<(), BoxError> {
+        let sender = create_sender_context()?;
+        let serialized = serde_json::to_string(&sender)?;
         let deserialized = serde_json::from_str(&serialized)?;
-        assert!(req_ctx == deserialized);
+        assert!(sender == deserialized);
+        Ok(())
+    }
+
+    #[test]
+    fn test_serialize_v2() -> Result<(), BoxError> {
+        let sender = create_sender_context()?;
+        let body = serialize_v2_body(
+            &sender.v1.psbt,
+            sender.v1.disable_output_substitution,
+            sender.v1.fee_contribution,
+            sender.v1.min_fee_rate,
+        );
+        assert_eq!(body.as_ref().unwrap(), &<Vec<u8> as FromHex>::from_hex(SERIALIZED_BODY_V2)?,);
+        Ok(())
+    }
+
+    #[test]
+    fn test_extract_v2_success() -> Result<(), BoxError> {
+        let sender = create_sender_context()?;
+        let ohttp_relay = EXAMPLE_URL.clone();
+        let result = sender.extract_v2(ohttp_relay);
+        assert!(result.is_ok(), "Expected Ok result, got: {:#?}", result.err());
+        let (request, context) = result?;
+        assert!(!request.body.is_empty(), "Request body should not be empty");
+        assert_eq!(request.url, EXAMPLE_URL.clone());
+        assert_eq!(context.endpoint, sender.v1.endpoint);
+        assert_eq!(context.psbt_ctx.original_psbt, sender.v1.psbt);
+        Ok(())
+    }
+
+    #[test]
+    fn test_extract_v2_fails_missing_pubkey() -> Result<(), BoxError> {
+        let expected_error = "cannot parse receiver public key: receiver public key is missing";
```

Comment:
Why not compare the Error directly instead of serializing it into a string?

Reply:
because the variant isn't `pub`